File: tests/test_opaque_types.py
from __future__ import annotations

import pytest

import env
from pybind11_tests import ConstructorStats, UserType
from pybind11_tests import opaque_types as m


def test_string_list():
    lst = m.StringList()
    lst.push_back("Element 1")
    lst.push_back("Element 2")
    assert m.print_opaque_list(lst) == "Opaque list: [Element 1, Element 2]"
    assert lst.back() == "Element 2"

    for i, k in enumerate(lst, start=1):
        assert k == f"Element {i}"
    lst.pop_back()
    assert m.print_opaque_list(lst) == "Opaque list: [Element 1]"

    cvp = m.ClassWithSTLVecProperty()
    assert m.print_opaque_list(cvp.stringList) == "Opaque list: []"

    cvp.stringList = lst
    cvp.stringList.push_back("Element 3")
    assert m.print_opaque_list(cvp.stringList) == "Opaque list: [Element 1, Element 3]"


def test_pointers(msg, backport_typehints):
    living_before = ConstructorStats.get(UserType).alive()
    assert m.get_void_ptr_value(m.return_void_ptr()) == 0x1234
    assert m.get_void_ptr_value(UserType())  # Should also work for other C++ types

    if not env.GRAALPY:
        assert ConstructorStats.get(UserType).alive() == living_before

    with pytest.raises(TypeError) as excinfo:
        m.get_void_ptr_value([1, 2, 3])  # This should not work

    assert (
        backport_typehints(msg(excinfo.value))
        == """
            get_void_ptr_value(): incompatible function arguments. The following argument types are supported:
                1. (arg0: types.CapsuleType) -> int

            Invoked with: [1, 2, 3]
        """
    )

    assert m.return_null_str() is None
    assert m.get_null_str_value(m.return_null_str()) is not None

    ptr = m.return_unique_ptr()
    assert "StringList" in repr(ptr)
    assert m.print_opaque_list(ptr) == "Opaque list: [some value]"


def test_unions():
    int_float_union = m.IntFloat()
    int_float_union.i = 42
    assert int_float_union.i == 42
    int_float_union.f = 3.0
    assert int_float_union.f == 3.0


def test_issue_5988_opaque_std_array():
    """Regression test for GitHub issue #5988: crash when binding with opaque std::array types."""
    # Test basic Array3d (opaque std::array<double, 3>) functionality
    a = m.Array3d()
    a[0] = 1.0
    a[1] = 2.5
    a[2] = 3.0
    assert a[0] == 1.0
    assert a[1] == 2.5
    assert a[2] == 3.0
    assert len(a) == 3

    with pytest.raises(IndexError):
        _ = a[3]

    # Test VecArray3d (opaque std::vector<std::array<double, 3>>) functionality
    v = m.VecArray3d()
    assert len(v) == 0
    v.push_back(a)
    assert len(v) == 1
    assert v[0][0] == 1.0
    assert v[0][1] == 2.5
    assert v[0][2] == 3.0

    with pytest.raises(IndexError):
        _ = v[1]

File: tests/test_opaque_types.cpp
/*
    tests/test_opaque_types.cpp -- opaque types, passing void pointers

    Copyright (c) 2016 Wenzel Jakob <wenzel.jakob@epfl.ch>

    All rights reserved. Use of this source code is governed by a
    BSD-style license that can be found in the LICENSE file.
*/

#include <pybind11/stl.h>

#include "pybind11_tests.h"

#include <array>
#include <vector>

// IMPORTANT: Disable internal pybind11 translation mechanisms for STL data structures
//
// This also deliberately doesn't use the below StringList type alias to test
// that MAKE_OPAQUE can handle a type containing a `,`.  (The `std::allocator`
// bit is just the default `std::vector` allocator).
PYBIND11_MAKE_OPAQUE(std::vector<std::string, std::allocator<std::string>>)

// Test for GitHub issue #5988: PYBIND11_MAKE_OPAQUE with std::array types.
// These types are not used as converted types in other test files, so they
// can safely be made opaque here without ODR violations.
PYBIND11_MAKE_OPAQUE(std::array<double, 3>)
PYBIND11_MAKE_OPAQUE(std::vector<std::array<double, 3>>)

using StringList = std::vector<std::string, std::allocator<std::string>>;

// Type aliases for issue #5988 test
using Array3d = std::array<double, 3>;
using VecArray3d = std::vector<Array3d>;

TEST_SUBMODULE(opaque_types, m) {
    // test_string_list
    py::class_<StringList>(m, "StringList")
        .def(py::init<>())
        .def("pop_back", &StringList::pop_back)
        /* There are multiple versions of push_back(), etc. Select the right ones. */
        .def("push_back", (void (StringList::*)(const std::string &)) &StringList::push_back)
        .def("back", (std::string & (StringList::*) ()) & StringList::back)
        .def("__len__", [](const StringList &v) { return v.size(); })
        .def(
            "__iter__",
            [](StringList &v) { return py::make_iterator(v.begin(), v.end()); },
            py::keep_alive<0, 1>());

    class ClassWithSTLVecProperty {
    public:
        StringList stringList;
    };
    py::class_<ClassWithSTLVecProperty>(m, "ClassWithSTLVecProperty")
        .def(py::init<>())
        .def_readwrite("stringList", &ClassWithSTLVecProperty::stringList);

    m.def("print_opaque_list", [](const StringList &l) {
        std::string ret = "Opaque list: [";
        bool first = true;
        for (const auto &entry : l) {
            if (!first) {
                ret += ", ";
            }
            ret += entry;
            first = false;
        }
        return ret + "]";
    });

    // test_pointers
    m.def("return_void_ptr", []() { return (void *) 0x1234; });
    m.def("get_void_ptr_value", [](void *ptr) { return reinterpret_cast<std::intptr_t>(ptr); });
    m.def("return_null_str", []() { return (char *) nullptr; });
    m.def("get_null_str_value", [](char *ptr) { return reinterpret_cast<std::intptr_t>(ptr); });

    m.def("return_unique_ptr", []() -> std::unique_ptr<StringList> {
        auto *result = new StringList();
        result->emplace_back("some value");
        return std::unique_ptr<StringList>(result);
    });

    // test unions
    py::class_<IntFloat>(m, "IntFloat")
        .def(py::init<>())
        .def_readwrite("i", &IntFloat::i)
        .def_readwrite("f", &IntFloat::f);

    // test_issue_5988: PYBIND11_MAKE_OPAQUE with std::array and nested containers
    // (Regression test for crash when importing modules with opaque std::array types)
    py::class_<Array3d>(m, "Array3d")
        .def(py::init<>())
        .def("__getitem__",
             [](const Array3d &a, std::size_t i) -> double {
                 if (i >= a.size()) {
                     throw py::index_error();
                 }
                 return a[i];
             })
        .def("__setitem__",
             [](Array3d &a, std::size_t i, double v) {
                 if (i >= a.size()) {
                     throw py::index_error();
                 }
                 a[i] = v;
             })
        .def("__len__", [](const Array3d &a) { return a.size(); });

    py::class_<VecArray3d>(m, "VecArray3d")
        .def(py::init<>())
        .def("push_back", [](VecArray3d &v, const Array3d &a) { v.push_back(a); })
        .def("__getitem__",
             [](const VecArray3d &v, std::size_t i) -> const Array3d & {
                 if (i >= v.size()) {
                     throw py::index_error();
                 }
                 return v[i];
             },
             py::return_value_policy::reference_internal)
        .def("__len__", [](const VecArray3d &v) { return v.size(); });
}
